frontend/lib/resolution/InitResolver.cpp

```diff
@@ -798,18 +835,36 @@ bool InitResolver::handleInitStatement(const uast::Init* node) {
 }
 
 bool InitResolver::handleUseOfField(const AstNode* node) {
-  auto id = fieldIdFromPossibleMentionOfField(node);
+  auto [id, isSuperField] = fieldIdFromPossibleMentionOfField(node);
   if (id.isEmpty()) return false;
 
-  if (isFieldInitialized(id)) return false;
+  if (!isSuperField && isFieldInitialized(id)) return false;
 
   auto state = fieldStateFromId(id);
   bool isValidPreInitMention = false;
 
-  if (isDescendingIntoAssignment_)
+  if (isDescendingIntoAssignment_ && isSuperField == false)
     if (isMentionOfNodeInLhsOfAssign(node))
       isValidPreInitMention = true;
 
+  // uses of parent fields might be valid during PHASE_NEED_COMPLETE if we
+  // implicitly resolved super.init(). We still need to keep track of parent
+  // field usage in case an explicit super.init() comes up later.
+  if (isSuperField) {
+    if (explicitSuperInit == false && phase_ != PHASE_COMPLETE) {
+      // Upon first use of parent field, need to insert a super.init() call
+      // TODO: store this as an associated action
+      if (useOfSuperFields_.empty()) {
+        this->resolveImplicitSuperInit();
+      }
+
+      useOfSuperFields_.push_back({id, node->id()});
+      return true;
+    } else {
```

Comment:
Do we need another case for when `explicitSuperInit == true` but the field usage comes before the `super.init` call?

I tried to test this out on the branch for myself but instead hit a segfault that might be indicative of another issue, with this code:
```
class A {
  var a, b : real;

  proc init(val : real) {
    this.a = val;
    this.b = val * 2;
  }
}

class B : A {
  var x, y : real;

  proc init(val: real, x: real, y: real) {
    this.a = val;
    super.init(val);
  }
}

var b1 = new B(4.0, 1.0, 2.0);
```

Reply:
I think the way things are set up, if ``explicitSuperInit`` is true it means that the field access is after the ``super.init()`` call.

The code example is running into a separate issue in which I/we need to improve error handling in the case where we either:
- cannot resolve an arg-less ``super.init()`` call
- or cannot resolve such a call when a valid ``super.init`` is called later

I'll make sure to open an issue to track this.

I had expected to be able to fix this by adding a default value for the argument of A's initializer, but apparently I overlooked the case where the user tried initializing a parent field explicitly. I have a fix for this and will add a new test soon.